Use dependency versions from the `grails-bom` where available

matrei opened the pull request:
And some other improvements and cleanup.

matrei committed:
test: Add log output to tests

matrei committed:
fix(deps): Use `grails-bom` versions where available

matrei committed:
chore: Improve comment in `documentation-config.gradle`

matrei committed:
test: Inject `JavaMailSender` to work with Spring 6

When creating the `JavaMailSender` via constructor, it could not connect to the greenmail server.

matrei committed:
chore: Cleanup `MailServiceSpec`

matrei committed:
Merge branch '5.0.x' into matrei/use-bom

matrei commented:
#56

sbglasius commented on build.gradle:
```diff
@@ -21,43 +21,47 @@ repositories {
-    implementation libs.javamail.impl
-    implementation libs.spring.beans
+    implementation libs.javamail.impl // Needs to be implementation (not runtimeOnly), as SMTPMessage is used in MailMessageBuilder
     implementation libs.springboot.autoconfigure
```
isn't this also in the grails bom?

matrei replied:
To my knowledge, spring boot is not in the `grails-bom`: https://repo.grails.org/ui/repos/tree/PomView/libs-snapshots-local/org/grails/grails-bom/7.0.0-SNAPSHOT/grails-bom-7.0.0-20241017.065954-109.pom

matrei replied:
The new `grails-bom` has the spring boot version. It is now used in this project.

sbglasius commented on build.gradle:
```diff
@@ -21,43 +21,47 @@ repositories {
 
-    integrationTestRuntimeOnly libs.grails.testing.support.web
-    integrationTestRuntimeOnly libs.slf4j.nop // Get rid of warning about missing slf4j implementation during test task
     integrationTestRuntimeOnly libs.springboot.starter.tomcat
```
and this too?

matrei replied:
To my knowledge, spring boot is not in the `grails-bom`: https://repo.grails.org/ui/repos/tree/PomView/libs-snapshots-local/org/grails/grails-bom/7.0.0-SNAPSHOT/grails-bom-7.0.0-20241017.065954-109.pom

matrei replied:
The new `grails-bom` has the spring boot version. It is now used in this project.

sbglasius reviewed (commented):
Two minor suggestions.

matrei committed:
fix(deps): Put all dependencies in version catalog

matrei committed:
fix(deps): Use more versions from `grails-bom`

`groovy-templates` and `grails-docs` are also in `grails-bom`

matrei committed:
fix(deps): Rename `grails-gsp` version catalog entry

matrei committed:
fix(deps): Set correct maven coords for `grails-gsp`

codeconsole commented on gradle/libs.versions.toml:
```diff
@@ -11,7 +11,7 @@ grails-bom = { module = 'org.grails:grails-bom', version.ref = 'grails' }
 greenmail = { module = 'com.icegreen:greenmail', version.ref = 'greenmail' }
 grails-docs = { module = 'org.grails:grails-docs' }
-grails-gsp = { module = 'org.grails.plugins:gsp' }
+grails-gsp = { module = 'org.grails:grails-gsp' }
```
Does the taglib @AstTransformer TagLibraryTransformer still get applied to the taglib?

matrei replied:
Good catch! I have added `org.grails:grails-web-taglib` to fix that.

codeconsole commented on gradle/libs.versions.toml:
```diff
@@ -1,36 +1,33 @@
-groovy = '4.0.23'
-gsp = '7.0.0-SNAPSHOT'
 javamail = '2.0.1'
 servlet-api = '6.0.0'
```
servlet api is in the bom

matrei replied:
> servlet api is in the bom

I think this comment might be on a previous version, this has been removed.
The only versions left are for `grails` and `greenmail`.

codeconsole replied:
I was just giving a general approve. That comment is from last week.

matrei committed:
fix(deps): Rename the aliases for build dependencies

To restrict the use of groups.

matrei committed:
fix(deps): Use all available versions from `grails-bom`

- Remove explicit versions to where versions are available in `grails-bom`
- Use `org.eclipse.angus:jakarta.mail` instead of `com.sun.mail:jakarta.mail`
- Restrict the use of version catalog alias groups

matrei committed:
fix(deps): Update `com.icegreen:greenmail`

`2.0.1` -> `2.1.0`

codeconsole commented:
Well, I will ask the obvious.... now that we use the bom for versions, what is the point of using toml? doesn't it just make things more confusing and result in 2x lines of code? Does it provide anything of value?

matrei commented:
> Does it provide anything of value?

Using a version catalog has several advantages.
Most notably:
- Declare dependency coordinates and versions once, and in a centralized location (DRY).
- Less cluttered build files (subjective)

codeconsole commented:
> > Does it provide anything of value?
> 
> Using a version catalog has several advantages. Most notably:
> 
> * Declare dependency coordinates and versions once, and in a centralized location (DRY).
> * Less cluttered build files (subjective)

uh, it looks like you are declaring them twice
https://github.com/gpc/grails-mail/blob/c0a82ec33efbb51edc4293e6b3c6d2d6af9e87d9/build.gradle#L24-L54

https://github.com/gpc/grails-mail/blob/c0a82ec33efbb51edc4293e6b3c6d2d6af9e87d9/gradle/libs.versions.toml#L15-L36

now that we use a bom, you could completely delete `libs.versions.toml` and not add any additional lines of code. 
That equals 36 less lines of maintenance.

matrei commented:
First of all, you are quoting an old version before the new `grails-bom` was in effect.

Second, this is indeed DRY because the coords and versions are never repeated but can be and are used in multiple places.

codeconsole commented:
DRY?
```
grails-core = { module = 'org.grails:grails-core' }
compileOnly libs.grails.core 
```

```gradle
compileOnly 'org.grails:grails-core'
```

You shouldn't have open another file to look up coordinates, but this is just my opinion. Hopefully it will evolve to something better someday... 😄

matrei commented:
> You shouldn't have open another file to look up coordinates

Why do you need to see the coordinates? If you do, ctrl-click in Intellij to go right to the declaration.

> DRY?

compileOnly 'jakarta.servlet:jakarta.servlet-api' // Provided by the servlet container
testImplementation 'jakarta.servlet:jakarta.servlet-api'

codeconsole commented:
> > DRY?
> 
> compileOnly 'jakarta.servlet:jakarta.servlet-api' // Provided by the servlet container 
> testImplementation 'jakarta.servlet:jakarta.servlet-api'

compared to?  how is that handled any better with toml?   

I suppose if that happened a lot, I would just create another dependency scope 😉

matrei commented:
It appears we have different interpretations of DRY. My emphasis is on data and logic, whereas I perceive from this discussion that your focus is on the number of characters.

Another advantage of using a version catalog that I mentioned is the centralized location. `grails-mail` is a pretty small single-project build but it still has dependency declarations in two build files. By using a version catalog, it's easy to see and manage all dependencies used in the project. Intellij will also help with flagging unused dependencies in the toml file.

codeconsole commented:
> compileOnly 'jakarta.servlet:jakarta.servlet-api' // Provided by the servlet container
> testImplementation 'jakarta.servlet:jakarta.servlet-api'

Can you explain your example?  how does toml handle that any better?

matrei commented:
> Can you explain your example? how does toml handle that any better?

The example is the alternative version, when not using a version catalog. As `'jakarta.servlet:jakarta.servlet-api` is needed in both `compileOnly` and `testImplementation`, the maven coordinates must declared in two places. With a version catalog, the coordinates are only declared once.

> My version of DRY is not characters. It is avoiding repeating yourself, which in my opinion, toml is not avoiding when used with a bom.

From https://en.wikipedia.org/wiki/Don%27t_repeat_yourself:
> The DRY principle is stated as "Every piece of knowledge must have a single, unambiguous, authoritative representation within a system".

> Do you honestly think it does not add an unnecessary additional learning curve to the majority of gradle users?

As I said earlier, I think version catalogs have several advantages. Gradle changes, hopefully to the better, for every version it releases. We should embrace those improvements, and I think version catalogs are not a very hard concept to grasp. If you are a totally new Gradle user, maybe it is even easier? Also with modern IDEs, you get auto-completion, click-through, unused code highlighting etc.

My subjective view is also that it is easier to read good named aliases, rather than parsing maven coordinate/version "gibberish" when I scan build files.

codeconsole commented:
> My subjective view is also that it is easier to read good named aliases, rather than parsing maven coordinate/version "gibberish" when I scan build files.

```
compileOnly 'jakarta.servlet:jakarta.servlet-api' 
testImplementation 'jakarta.servlet:jakarta.servlet-api'
```
vs 2 files for:
```
servlet-api = { module = 'jakarta.servlet:jakarta.servlet-api' }

compileOnly libs.servlet.api 
testImplementation libs.servlet.api 
```
I honestly don't see the win here. Using toml you are still repeating yourself, and making it more difficult to follow.  Not everyone uses IntelliJ and IntelliJ click through doesn't always work.  The Grails IntelliJ plugin was broken for a year on Grails 6 and you couldn't click through on anything you used to be able to click on. In the end, the coordinates are the coordinates. Assigning a variable to them in a proprietary experimental format that has still yet to stabilize and provide a full feature set just seems like an over complication.

It's basically the route Grails 6 took over 5. It overcomplicated everything adding an unnecessary buildSrc folder and a bunch of configuration in settings.gradle.  It just made things more difficult to follow and confused a lot of people.  There are a lot of people thoroughly confused by this type of versioning as well. 

At the end of the day you have **TEN** build files (*.gradle +*.toml) for a very simple plugin which unfortunately will cause a lot of people to get lost who would like to contribute.   Why are we making things so complicated?  Convention over configuration and DRY are paradigms to make things simpler, not more complicated.  In my opinion, this plugin should consist of 1 build.gradle and that is it.  

We will have to agree to disagree here.  I am fine with the merge. I think you did a good job cleaning it up further. I just like to keep things as simple as possible and look fro ways to remove configuration, not add it.

matrei commented:
Agreed :wink: 

While the benefits in this project may seem marginal, I believe they become more apparent when these concepts are applied to more complex projects.
